user_mgr.cpp

```diff
@@ -1697,42 +1697,35 @@ std::vector<std::string> UserMgr::getFailedAttempt(const char* userName)
     return executeCmd("/usr/sbin/faillock", "--user", userName);
 }
 
-std::set<MultiFactorAuthType>& allAuthTypes()
-{
-    using namespace sdbusplus::common::xyz::openbmc_project::user::details;
-    static std::set<MultiFactorAuthType> authTypeSet;
-    if (authTypeSet.empty())
-    {
-        for (const auto& item : mappingMultiFactorAuthConfigurationType)
-        {
-            authTypeSet.insert(std::get<1>(item));
-        }
-    }
-    return authTypeSet;
-}
 MultiFactorAuthType UserMgr::enabled(MultiFactorAuthType value, bool skipSignal)
 {
     switch (value)
     {
         case MultiFactorAuthType::None:
-            for (auto type : allAuthTypes())
+            for (auto type : {MultiFactorAuthType::GoogleAuthenticator})
             {
                 for (auto& u : usersList)
                 {
                     u.second->enableMultiFactorAuth(type, false);
                 }
             }
             break;
-        case MultiFactorAuthType::GoogleAuthenticator:
+        default:
             for (auto& u : usersList)
             {
-                u.second->enableMultiFactorAuth(
-                    MultiFactorAuthType::GoogleAuthenticator, true);
+                u.second->enableMultiFactorAuth(value, true);
             }
             break;
     }
     return MultiFactorAuthConfigurationIface::enabled(value, skipSignal);
 }
-
+bool UserMgr::isGenerateSecretKeyRequired(std::string userName)
+{
+    if (usersList.contains(userName))
+    {
+        return usersList[userName]->secretKeyGenerationRequired();
+    }
```

Comment:
GenerateSecretKeyRequired D-bus property at user level may go away then we need to change this method.

Reply:
No It should not go away. Ideally it is property of user not usermanager. Otherwise you are breaking the encapsulation. In thiis case it jis ust that usr_mgr just relay on user level properties to return proper value. The real bug is with ldap implementation where there is no dbus representation for ldap user. Implementing usg_mgr method in terms of user level property is the right approach here.